### PR TITLE
Substitute REST path parameters in Workflow Steps

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

When provisioning a workflow (or creating a workflow with provision=true), additonal parameters can be passed on the REST path.  For example:
```
POST /_plugins/_flow_framework/workflow/bhPBgI0BQfVbnJKtHwGU/_provision?foo=one&bar=two
```

In this case, instances of the parameter in the template values (`user_params`, `user_inputs`, or `previous_node_inputs`) will be replaced by its value, e.g., 
```json
"user_inputs": {
  "parameters": {
    "math": "${{foo}} plus ${{ foo }} is ${{ bar }}"
  }
}
```
will become
```json
"user_inputs": {
  "parameters": {
    "math": "one plus one is two"
  }
}
```

### Issues Resolved

Part of #522 and a simplified/partial implementation of #496.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
@YANG-DB FYI

**Comment by dbwiddis** [comment]
Flipping it back to review. `WorkflowRequest` needs to be writeable and `WorkflowData` is not writeable because of the content map.  Implementation of #496 will require translating to a string-to-string map, and for that we can leverage the existing params map.

**Comment by owaiskazi19** [comment]
Continuing the discussion [here](https://github.com/opensearch-project/flow-framework/issues/213#issuecomment-1953022946)

**Review by joshpalis — APPROVED** [review]
Overall looks good to me, just a small comment

File: src/main/java/org/opensearch/flowframework/transport/ProvisionWorkflowTransportAction.java [hunk 1]
```diff
@@ -126,7 +126,11 @@ protected void doExecute(Task task, WorkflowRequest request, ActionListener<Work
 
                 // Sort and validate graph
                 Workflow provisionWorkflow = template.workflows().get(PROVISION_WORKFLOW);
-                List<ProcessNode> provisionProcessSequence = workflowProcessSorter.sortProcessNodes(provisionWorkflow, workflowId);
+                List<ProcessNode> provisionProcessSequence = workflowProcessSorter.sortProcessNodes(
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/ProvisionWorkflowTransportAction.java:129]: Good comment and I had that thought initially.  However, it may be that our particular wildcard substitution may conflict with some other template (it's the standard github action substitutions).   The various ML commons configurations do have nested JSON "templates" in some cases.   

We're protected in this case because we only substitute if the param is actually included when calling provision, so as long as there's not both a conflicting wildcard pattern AND same-named param it won't hurt.  But if we decide to try to validate/enforce a parameter because we see the pattern it could lead to problems.

[the remark this answers is not in the document]
Comment (joshpalis) [reply src/main/java/org/opensearch/flowframework/transport/ProvisionWorkflowTransportAction.java:129]: That makes sense, I didnt consider conflicting substitutions. In this case, I think we can rely on the underlying transport actions to validate the input values fully.

**Review by owaiskazi19 — COMMENTED** [review]
Overall LGTM with few questions/comments

**Comment by amitgalitz** [comment]
In the description you highlight the changes for provision API and mention when provision=true on create. If we are creating a template without the provision parameter(false) then should we also allow users to have dynamic params?

Additionally then for a larger template could I give this for the create operation and wouldn't we get validation failures:
```
{
  "name": "{template.name}",
  "description": "{template.descirption}",
  "use_case": "MODEL_DEPLOYMENT_API_KEY",
  "version": {
    "template": "1.0.0",
    "compatibility": [
      "2.12.0",
      "3.0.0"
    ]
  },
  "workflows": {
    "provision": {
      "nodes": [
        {
          "id": "create_connector_1",
          "type": "create_connector",
          "user_inputs": {
            "name": "${create_connector_1}",
            "description": "${create_connector_1description}",
            "version": "1",
            "protocol": "${create_connector_1.protocol}",
            "parameters": {
              "endpoint": "${create_connector_1.endpoint}",
              "model": "${create_connector_1.model}"
            },
            "credential": {
              "key": "${create_connector_1.credential.key}
            },
            "actions": [
              {
                "action_type": "predict",
                "method": "POST",
                "url": "https://${parameters.endpoint}/v1/chat/completions"
              }
            ]
          }
        },
        {
          "id": "register_remote_model_1",
          "type": "register_remote_model",
          "previous_node_inputs": {
            "create_connector_step_1": "parameters"
          },
          "user_inputs": {
            "name": "${register_remote_1_model.name}",
            "function_name": "remote",
            "description": "${register_remote_model_1.description}"
          }
        },
         {
            "id": "deploy_model_1",
            "type": "deploy_model",
            "previous_node_inputs": {
              "register_model_1": "model_id"
            }
          }
      ],
      "edges": [
        {
          "source": "create_connector_1",
          "dest": "register_model_1"
        },
         {
            "source": "register_model_1",
            "dest": "deploy_model_1"
          }
      ]
    }
  }
}
```

File: src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java [hunk 2]
```diff
@@ -84,12 +100,24 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
                 new BytesRestResponse(ffe.getRestStatus(), ffe.toXContent(channel.newErrorBuilder(), ToXContent.EMPTY_PARAMS))
             );
         }
+        if (!provision && !params.isEmpty()) {
+            // Consume params and content so custom exception is processed
+            params.keySet().stream().forEach(request::param);
+            request.content();
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java:106]: The `BaseRestHandler` tracks whether parameters and content (body) are "consumed" by the RestHandler implementations.  The apparent intent of the code is to provide for some sort of "did you mean?" functionality if someone made a typo declaring a parameter.  (If you included the parameter `food` but never consumed `foo` it could be suggested as a replacement.)

Unfortunately, this check for consumed parameters overrides the returned value from the channel.  See [here](https://github.com/opensearch-project/OpenSearch/blob/247e2eec46930e2eaec970b1acfb2f7d18e3e7a9/server/src/main/java/org/opensearch/rest/BaseRestHandler.java#L106-L129) where the `action` returned from from the `prepareRequest()` method in line 1-6 isn't actually returned to the channel until line 109, while the lack of consuming params and content can throw exceptions before it gets there.  

I've known about this issue since I had to handle it in the SDK, but it reared its ugly head checking the feature flag in this repo, giving misleading error messages.  See https://github.com/opensearch-project/flow-framework/issues/513

TLDR: if you have any params in the rest path, or parse the content, and if you want your own custom message returned, you have to "consume" the params (by calling `request.param("foo")` and `content()`) before returning.

File: src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java [hunk 3]
```diff
@@ -68,24 +73,43 @@ public List<Route> routes() {
...
+            // If body is included get any params from body
+            if (request.hasContent()) {
+                try (XContentParser parser = request.contentParser()) {
+                    ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                    while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        String key = parser.currentName();
+                        if (params.containsKey(key)) {
+                            throw new FlowFrameworkException("Duplicate key " + key, RestStatus.BAD_REQUEST);
+                        }
+                        if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java:92]: Because they are eventually passed to the `ProvisionWorkflowTransportAction` in the `WorkflowRequest`.  Since it uses the transport protocol, it uses `Writeable` implementation (`writeTo()` and `StreamInput` constructor) to pass the parameter map to the provisioning action.  

In theory we could do any `Writeable` implementation and the streams [do detect a finite number of simple subtitutions](https://github.com/opensearch-project/OpenSearch/blob/247e2eec46930e2eaec970b1acfb2f7d18e3e7a9/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamOutput.java#L795-L812) but that's really hard to try to document to users (even I don't understand it fully).  

We can always start with String and make it more flexible in the future if we get requests, but it's too hard to cover (and document) all the nuances of what is and is not writeable.  It's why JSON was invented and, fun enough, JSON is a string, which is where we'd end up parsing this on the other end anyway!  That's how we are returning the `Template` object in the get workflow API because making that whole object `Writable` is a lot harder than just sending a string over transport.

File: src/main/java/org/opensearch/flowframework/transport/WorkflowRequest.java [hunk 4]
```diff
@@ -77,6 +105,7 @@ public WorkflowRequest(StreamInput in) throws IOException {
         this.template = templateJson == null ? null : Template.parse(templateJson);
         this.validation = in.readStringArray();
         this.provision = in.readBoolean();
+        this.params = this.provision ? in.readMap(StreamInput::readString, StreamInput::readString) : Collections.emptyMap();
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/WorkflowRequest.java:108]: The `StreamInput` constructor implementation must exactly match the fields written in `writeTo()` where we only write the map in the case `provision` is true:
```java
        if (provision) {
            out.writeMap(params, StreamOutput::writeString, StreamOutput::writeString);
        }
```
We could have used the "optional map" features here but that would have just created a redundant boolean byte in the transport serialization, or just always written the map without that conditional which would have written a variable integer 0 byte for size.  Either way that's an extra byte.

If we're not provisioning, the map is defined to be (and throws an exception if it is not) empty.

**Comment by dbwiddis** [comment]
> If we are creating a template without the provision parameter(false) then should we also allow users to have dynamic params?

No, because the params are only substituted during provisioning, when processing the data in the Workflow Steps.  They have no impact when we are not provisioning, as we are simply parsing and storing the template content.

> Additionally just so I understand correctly, technically a user can create this with create API and then when provisioning give the input to the params or give it during provisioning?

Yes, classic example is the feature request issue.  The template would include a step configuration with something like this (this exact syntax was from observability workflow RFC, our syntax is slightly different).
```
"steps": [
{
  "name": "add_timestamp_field_to_otel_spans_index",
  "method": "PUT",
  "endpoint": "/${index_name}",
  "body": { ... }
 },
 ....
]
```
In this case we'd create the workflow with the "endpoint" field having a value of `/${{ index_name }}`.

When we went to provision the workflow, if we passed `?index_name=foo` on the REST path, then the "endpoint" field value would become `/foo`.  The original template would remain unchanged.

Note that in our current code we couldn't provision a second time in this case, we'd have to "deprovision" first and I'll have to eventually address this before 2.13.  We could then deprovision (TODO, not require this) and provision again passing `?index_name=bar` on the REST path.  Same template, but "endpoint" would now be `/bar`.

File: src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java [hunk 5]
```diff
@@ -68,24 +73,43 @@ public List<Route> routes() {
...
+                            throw new FlowFrameworkException("Request body fields must have string values", RestStatus.BAD_REQUEST);
+                        }
+                        params.put(key, parser.text());
+                    }
+                } catch (IOException e) {
+                    throw new FlowFrameworkException("Request body parsing failed", RestStatus.BAD_REQUEST);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java]: Done.

File: src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java:106]: On the same line, looking at [this](https://github.com/opensearch-project/flow-framework/blob/main/src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java#L98), we are returning a ffe and we do have params in the rest path. Does calling `request.param` [here](https://github.com/opensearch-project/flow-framework/blob/main/src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java#L70), helped to consume it or we need to call `request.content()` as well?

File: src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java]: > We can always start with String and make it more flexible in the future if we get requests,

It's fine to start with String and then increment later. For now, do you also want to include type `VALUE_BOOLEAN`?

File: src/main/java/org/opensearch/flowframework/transport/WorkflowRequest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/transport/WorkflowRequest.java:108]: That makes sense. I was thinking in terms of mutability. Generally, when the map has some entries, using `readMap`can make it as mutable and when it doesn't have anything it can be immutable in nature.

File: src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java:106]: > On the same line, looking at [this](https://github.com/opensearch-project/flow-framework/blob/main/src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java#L98), we are returning a ffe and we do have params in the rest path. Does calling `request.param` [here](https://github.com/opensearch-project/flow-framework/blob/main/src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java#L70), helped to consume it 

Precisely.  Currently (before this PR) the only allowable (and required) parameter for the provision API is the Workflow ID.  We "consume" the workflow ID param by calling `request.param(WORKFLOW_ID)`.  You'll notice ALL our API steps do that (see #513).  Some have a few params we need to consume.

> or we need to call `request.content()` as well?

This is similar, but just a boolean.  Prior to this PR, the provision API did not support having a body so we could rely on the `BaseRestHandler` error message, but had to at least return _something_, thus we send "null" back to the channel which gets ignored when the no-content-consumed exception is thrown.

File: src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java]: > For now, do you also want to include type VALUE_BOOLEAN?

No, because then I'd have to modify the `Map<String,String>` currently passed in the `WorkflowRequest` to a `Map<String, SomeOtherClassThatCanBeBothStringAndBoolean>`.  Which is pretty much just `Object`. :)

If I expanded it at all I'd use the same structure as the Settings API for a reference.   There are individual `Setting<T>` objects of various types, but the `Settings` class itself maintains a `String, String` map, or a `String, Object` map where the only acceptable object is a nested map.   The `Setting` objects themselves do the work of parsing themselves when you extract the (string) value on the other end of transport.

So the choices are:
1. Just specify a string, and deal with all the type parsing internally in the appropriate workflow step, e.g., specify port as "1234" and parse it to an int inside the step (which is what I do in the HttpHost step).  So if a user wants a boolean they just put "true".
2. Do a whole bunch of parsing before serializing for transport which just turns it into strings anyway and then parses it out on the other side, but we'd have to add a lot more overhead on transport to tell it HOW to parse.

File: src/main/java/org/opensearch/flowframework/transport/WorkflowRequest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/WorkflowRequest.java:108]: > That makes sense. I was thinking in terms of mutability. Generally, when the map has some entries, using `readMap`can make it as mutable and when it doesn't have anything it can be immutable in nature.

You can do `HashMap::new` if you want a mutable map here.

**Comment by amitgalitz** [comment]
> Note that in our current code we couldn't provision a second time in this case, we'd have to "deprovision" first and I'll have to eventually address this before 2.13. We could then deprovision (TODO, not require this) and provision again passing `?index_name=bar` on the REST path. Same template, but "endpoint" would now be `/bar`.

Makes sense to me, out of scope of this PR we could potentially in 2.13 add deprovision as a parameter or try to solve it with the versioning param. 

Could we also potentially add an ability to reprovision a workflow and give it a new ID for the state index breaking the 1:1 template to state index ratio. Up to customer to change things if there are issues.

File: src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/rest/RestProvisionWorkflowAction.java]: Dealing with parsing internally sounds good.

**Review by owaiskazi19 — APPROVED** [review]
LGTM! Thanks for being patient with the review and comments
